### ISSUE TITLE
Reduction types

### ISSUE THREAD

**Issue by jamesaross** [openshmem-org/specification#235]
OpenSHMEM reductions have a mixed bag of type support. For reference:
AND, OR, and XOR: `short, int, long, long long`
MAX and MIN: `short, int, double, float, long, long double, long long`
SUM and PROD: `double _Complex, float _Complex, short, int, double, float, long, long double, long long`

Obviously, not all operations of certain types make logical sense or have order of operations issues.  But a large number of types were added for RMA and atomics in 1.4 for specification completeness, whereas the reductions part of the specification was neglected.  Some RMA and atomic typed operations are of dubious utility but they were added nonetheless.

I don't recall the original rationale for reduction type selection.  Whereas atomic operations often have hardware support for specific data types, reduction operations are most likely software based in most implementations, which is why it seems counter-intuitive to me to have better type support for atomics and RMA vs. reductions.

Additionally, the reduction operations didn't get wrapped up in C11 Generics like much of the rest of the specification.

Would it make sense to include all of the integral types defined in Table 3 for RMA operations? For reference:
`char, signed char, short, int, long, long long, unsigned char, unsigned short, unsigned int, unsigned long, unsigned long long, int8_t, int16_t, int32_t, int64_t, uint8_t, uint16_t, uint32_t, uint64_t, size_t, 
ptrdiff_t`

In other words, should I put together a pull request with the following?
- AND, OR, and XOR would support all integral types and equivalent C11 Generics
- MAX and MIN would support `double, float`, and all integral types and equivalent C11 Generics
- SUM and PROD would support `double _Complex, float _Complex, double, float`, and all integral types and equivalent C11 Generics

Also, that other API that people sometimes use supports Logical AND (`&&`) and Logical OR (`||`) reductions operations.  Is there any interest in supporting those operations?

**Comment by jdinan** [comment]
We have not updated the collectives API because we anticipate adding new collectives interfaces as part of the teams proposal.  I would suggest joining the teams WG (see wiki for meeting info) and merging these good ideas into that proposal.

**Comment by jdinan** [comment]
@jamesaross Is this resolved by the new teams collectives API?

**Comment by jdinan** [comment]
@jamesaross Ping

**Comment by jamesaross** [comment]
@jdinan No, this wasn't resolved.  Looks like it was pushed to 1.6.

**Comment by jdinan** [comment]
IIRC, we decided that we would not introduce any new types for the legacy (active set) collectives. Does this issue exist with the teams collectives? If so, we should try to address it in OpenSHMEM 1.5.

**Comment by jamesaross** [comment]
@jdinan, It depends which teams collectives you're discussing.  If you're discussing the `shmem_collect` and `shmem_fcollect` team collectives, they currently use Table 5 types in the 1.5rc1 draft, which include all of the integral types (see Table 5 below).  If you mean the `shmem_[OP]_reduce` collectives, they use Table 9, which doesn't list most integral types (see Table 9 below).

 * I think it makes sense to expand for new team reductions in 1.5, but I don't feel strongly about it either way
 * I don't think it makes sense to expand list of deprecated active set reductions

I had forgotten I brought this issue up again in #279 and @nspark has [an improvement which looks good (except for `short *pWrk` on active set reductions](https://github.com/nspark/specification/blob/e12d71e8987079631786154e7af9b83f67110d74/content/shmem_reductions.tex).  It's not a single commit and I'm not sure how to have him submit a pull request. Separately, I submitted a pull request #392 to fix the deprecated reduction `short *pWrk` type.

### Table 5 type support for team `shmem_collect`/`shmem_fcollect`:
| **TYPE** | **TYPENAME** |
|----|----|
| float | float |
| double | double |
| long double | longdouble |
| char | char |
| signed char | schar |
| short | short |
| int | int |
| long | long |
| long long | longlong |
| unsigned char | uchar |
| unsigned short | ushort |
| unsigned int | uint |
| unsigned long | ulong |
| unsigned long long | ulonglong |
| int8_t | int8 |
| int16_t | int16 |
| int32_t | int32 |
| int64_t | int64 |
| uint8_t | uint8 |
| uint16_t | uint16 |
| uint32_t | uint32 |
| uint64_t | uint64 |
| size_t | size |
| ptrdiff_t | ptrdiff |


### Table 9 type support for team reductions (`shmem_[TYPENAME]_[OP]_reduce`):
| **TYPE** | **TYPENAME** | (*Operations* | *Supporting* | *TYPE*) |
|------|------|------|------|------|
| unsigned char | uchar | AND, OR, XOR | | |
| short | short | AND, OR, XOR | MAX, MIN | SUM, PROD |
| unsigned short | ushort | AND, OR, XOR | MAX, MIN | SUM, PROD |
| int | int | AND, OR, XOR | MAX, MIN | SUM, PROD |
| unsigned int | uint | AND, OR, XOR | MAX, MIN | SUM, PROD |
| long | long | AND, OR, XOR | MAX, MIN | SUM, PROD |
| unsigned long | ulong | AND, OR, XOR | MAX, MIN | SUM, PROD |
| long long | longlong | AND, OR, XOR | MAX, MIN | SUM, PROD |
| unsigned long long | ulonglong | AND, OR, XOR | MAX, MIN | SUM, PROD |
| float | float | | MAX, MIN | SUM, PROD |
| double | double | | MAX, MIN | SUM, PROD |
| long double | longdouble | | MAX, MIN | SUM, PROD |
| double _Complex | complexd | | | SUM, PROD |
| float _Complex | complexf | | | SUM, PROD |
